File: src/13-B-water_quality.R
library(tidyverse)
library(data.table)
library(arrow)

TABLE_PASSAGE_DF <- read_feather("results_building/t_passage.feather") |>
  select(code_insee24, arr24, bv2022) |>
  unique()

COMS <- fread("data/external/msp-eau/DIS_COM_UDI_2023.txt") |>
  select(cdreseau, inseecommune) |>
  unique()

PLV <- fread("data/external/msp-eau/DIS_PLV_2023.txt") |>
  select(referenceprel, cdreseau) |>
  unique() |>
  left_join(COMS, by= "cdreseau") |>
  select(referenceprel, code_insee24 = inseecommune) |>
  unique()

WATER_Q <- fread("data/external/msp-eau/DIS_RESULT_2023.txt")

WATER_Q <- WATER_Q[cdparametresiseeaux%in%c("CL", "NO3", "PH", "SO4")] 

WATER_Q[, "rqana" := as.numeric(str_replace(str_remove(rqana, "<|>"), ",", "."))] 

WATER_Q <- WATER_Q[, .(rqana = mean(rqana, na.rm = TRUE)), by = .(referenceprel, cdparametresiseeaux)]

WATER_Q <- WATER_Q |>
  filter(!is.na(rqana)) |>
  select(referenceprel, water_param = cdparametresiseeaux, rqana) 

WATER_QB <- WATER_Q |> 
  left_join(PLV, by = "referenceprel") |>
  group_by(code_insee24, water_param) |>
  summarise(rqana = mean(rqana)) |>
  inner_join(TABLE_PASSAGE_DF, by= "code_insee24")

sum_water <- function(DATA, level){
  DATA |>
    group_by({{level}}, water_param) |>
    summarise(rqana = mean(rqana, na.rm=T)) |>
    pivot_wider(names_from = water_param, values_from = rqana) |>
    rename_with( ~ paste0("water_",tolower(.x)), -{{level}})
}

WATERQ_DEP <- sum_water(WATER_QB |> mutate(dep = substr(code_insee24, 1, 2)), dep)

WATERQ_ARR <- sum_water(WATER_QB, arr24)

WATERQ_BV22 <- sum_water(WATER_QB, bv2022) |>
  filter(!(is.na(water_cl)|is.na(water_no3)|is.na(water_ph)|is.na(water_so4)))

write_feather(as.data.frame(WATERQ_ARR), "data/interim/arrondissements/water_quality.feather", compression = "zstd")
write_feather(as.data.frame(WATERQ_BV22), "data/interim/bv2022/water_quality.feather", compression = "zstd")
write_feather(as.data.frame(WATERQ_DEP), "data/interim/departement/water_quality.feather", compression = "zstd")

File: src/13-B-water_quality.py
import pandas as pd

# Import Data -------------
table_passage_df = (
    pd.read_feather("results_building/t_passage.feather")[[
        "code_insee24", "arr24", "bv2022"
    ]]
    .drop_duplicates()
)

coms = (
    pd.read_csv(
        "data/external/msp-eau/DIS_COM_UDI_2023.txt",
        sep=",",
        dtype=str,
        low_memory=False
    )
    [["cdreseau", "inseecommune"]]
    .drop_duplicates()
)

plv = (
    pd.read_csv(
        "data/external/msp-eau/DIS_PLV_2023.txt",
        sep=",",
        dtype=str,
        low_memory=False
    )
    [["referenceprel", "cdreseau"]]
    .drop_duplicates()
    .merge(coms, on="cdreseau", how="left")
    [["referenceprel", "inseecommune"]]
    .rename(columns={"inseecommune": "code_insee24"})
    .drop_duplicates()
)

# Prepare Water Quality DF -------------------
water_q = pd.read_csv(
    "data/external/msp-eau/DIS_RESULT_2023.txt",
    sep=",",
    dtype=str,
    low_memory=False
)

water_q = water_q[water_q["cdparametresiseeaux"].isin(["CL", "NO3", "PH", "SO4"])].copy()

water_q["rqana"] = (
    water_q["rqana"]
      .str.replace(r"[<>]", "", regex=True)
      .str.replace(",", ".", regex=False)
)
water_q["rqana"] = pd.to_numeric(water_q["rqana"], errors="coerce")

water_q = (
    water_q
    .groupby(["referenceprel", "cdparametresiseeaux"], as_index=False)["rqana"]
    .mean()
)
water_q = water_q.dropna(subset=["rqana"])
water_q = water_q.rename(columns={"cdparametresiseeaux": "water_param"})

water_qb = (
    water_q
    .merge(plv, on="referenceprel", how="left")
    .groupby(["code_insee24", "water_param"], as_index=False)["rqana"]
    .mean()
)

water_qb = water_qb.merge(table_passage_df, on="code_insee24", how="inner")

def sum_water(df: pd.DataFrame, level_col: str) -> pd.DataFrame:
    grp = (
        df.groupby([level_col, "water_param"], as_index=False)["rqana"]
        .mean()
    )
    piv = grp.pivot(index=level_col, columns="water_param", values="rqana").reset_index()
    piv.columns.name = None
    # rename the parameter columns to water_<param_in_lowercase>
    rename_map = {
        col: f"water_{col.lower()}" for col in piv.columns if col != level_col
    }
    return piv.rename(columns=rename_map)

# Main logic -------------------------
waterq_dep = sum_water(
    water_qb.assign(dep=water_qb["code_insee24"].str[:2]),
    "dep"
)

waterq_arr = sum_water(water_qb, "arr24")

waterq_bv22 = (
    sum_water(water_qb, "bv2022")
    .dropna(subset=["water_cl", "water_no3", "water_ph", "water_so4"])
)

# Export ------------------------
waterq_arr.to_feather(
    "data/interim/arrondissements/water_quality.feather",
    compression="zstd"
)
waterq_bv22.to_feather(
    "data/interim/bv2022/water_quality.feather",
    compression="zstd"
)
waterq_dep.to_feather(
    "data/interim/departement/water_quality.feather",
    compression="zstd"
)
